sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -0,0 +1,380 @@
+  @transient private val queryStagePreparationRules: Seq[Rule[SparkPlan]] = Seq(
+    PlanAdaptiveSubqueries(subqueryMap),
+    EnsureRequirements(conf)
+  )
+
+  // A list of physical optimizer rules to be applied to a new stage before its execution. These
+  // optimizations should be stage-independent.
+  @transient private val queryStageOptimizerRules: Seq[Rule[SparkPlan]] = Seq(
+    CollapseCodegenStages(conf)
+  )
+
+  private var currentStageId = 0
+
+  @volatile private var currentPhysicalPlan =
+    applyPhysicalRules(initialPlan, queryStagePreparationRules)
+
+  /**
+   * Return type for `createQueryStages`
+   * @param newPlan the new plan with created query stages.
+   * @param allChildStagesMaterialized whether all child stages have been materialized.
+   * @param newStages the newly created query stages, including new reused query stages.
+   */
+  private case class CreateStageResult(
+    newPlan: SparkPlan,
+    allChildStagesMaterialized: Boolean,
+    newStages: Seq[(Exchange, QueryStageExec)])
+
+  /**
+   * This method is called recursively to traverse the plan tree bottom-up and checks whether or
+   * not the child query stages (if any) of the current node have all been materialized. If the
+   * current node is:
+   * 1. An Exchange:
+   *    1) If the Exchange is equivalent to another Exchange of an existing query stage, create
+   *       a reused query stage and replace the current node with the reused stage.
+   *    2) If all its child query stages have been materialized, create a new query stage and
+   *       replace the current node with the new stage.
+   *    3) Otherwise, do nothing.
+   *    In any of the above cases, the method will return `false` for `allChildStagesMaterialized`.
+   * 2. A query stage:
+   *    return `true` for `allChildStagesMaterialized` if this stage is completed, otherwise
+   *    `false`.
+   * 3. Recursively visit all child nodes, and propagate the return value.
+   */
+  private def createQueryStages(plan: SparkPlan): CreateStageResult = plan match {
+    case e: Exchange =>
+      stageCache.get(e.canonicalized) match {
+        case Some(existingStage) if conf.exchangeReuseEnabled =>
+          val reusedStage = reuseQueryStage(existingStage, e.output)
+          CreateStageResult(newPlan = reusedStage,
+            allChildStagesMaterialized = false, newStages = Seq((e, reusedStage)))
+
+        case _ =>
+          val result = createQueryStages(e.child)
+          val newPlan = e.withNewChildren(Seq(result.newPlan)).asInstanceOf[Exchange]
+          // Create a query stage only when all the child query stages are ready.
+          if (result.allChildStagesMaterialized) {
+            var newStage = newQueryStage(newPlan)
+            if (conf.exchangeReuseEnabled) {
+              val queryStage = stageCache.getOrElseUpdate(e.canonicalized, newStage)
+              if (queryStage.ne(newStage)) {
+                newStage = reuseQueryStage(queryStage, e.output)
```

Comment:
wait, the `stageCache` is already updated, is it safe to reuse the query stage that has been removed from `stageCache`?

Reply:
This is to try reusing an existing again. And it's for avoiding a big lock on the entire "create-or-reuse-stage" block. It works like this:
1. We first try "getting" an existing stage for reuse as soon as we see an `Exchange` without going into the node and see if there's any un-materialized child stages. If there's a match, we are happy and will reuse it immediately.
2. If there isn't a match, we keep traversing into the `Exchange` and see if all child stages are ready. If not, we'll just hold back from creating a new stage.
3. If yes, that means we can create a new stage on this `Exchange` now, but there are chances that between step (1) and now, there might be a new stage that can match this `Exchange`. So we do a `getOrElseUpdate` and there can be 2 outcomes:
  a) the returned stage is exactly the new stage used for "ElseUpdate", then we just go ahead with the new stage.
  b) otherwise, it's an existing stage that gets put into the `stageCache` between step (1) and now, so let's forget about the new stage we've just created and reuse the existing stage we've found.